### ISSUE TITLE
Support Queues

### ISSUE THREAD

**Issue by Atinux** [nuxt-hub/core#135]
See https://developers.cloudflare.com/queues/

How would you see the usage within NuxtHub? Feedback welcome ❤️

**Comment by RihanArfan** [comment]
Consumers could be defined similar to other Nitro events in `server/queues/consumer-example.ts`. Would probs be like Nitro's event object syntax so that [details to configure it](https://developers.cloudflare.com/queues/configuration/configure-queues/#workers) can be done alongside the code. 

For producers, a composable like [`hubQueues().send()`](https://developers.cloudflare.com/queues/configuration/javascript-apis/#producer) would suffice. Could have some fancy DX around autocompletion for queue bindings both in consumer objects and producer function.

It's worth noting that Cloudflare are introducing a [thing for workflows](https://blog.cloudflare.com/data-anywhere-events-pipelines-durable-execution-workflows) which I think should be considered alongside implementing queues as it may be better long term for handling complex applications better? See [this Tweet](https://x.com/elithrar/status/1798397582739406876) and [this blog post](https://blog.cloudflare.com/data-anywhere-events-pipelines-durable-execution-workflows).

**Comment by Atinux** [comment]
> Consumers could be defined similar to other Nitro events in server/queues/consumer-example.ts. Would probs be like Nitro's event object syntax so that details to configure it can be done alongside the code.

I really like this idea!

Regarding the workflow, would you mind creating a new issue for it and linking this one too?

**Comment by Gerbuuun** [comment]
I'm not sure how this works with pages? Currently I have a separate worker project which is the queue consumer.
Does pages support this or is it only pull based (http)?

It would be nice to be able to merge this consumer code back into my main Nuxt project!

**Comment by Atinux** [comment]
I am not sure that Pages support queues but I know it's planned.

I try to avoid having to create two outputs and 2 deployments for one project.

**Comment by RihanArfan** [comment]
> I'm not sure how this works with pages? Currently I have a separate worker project which is the queue consumer. Does pages support this or is it only pull based (http)?
> 
> It would be nice to be able to merge this consumer code back into my main Nuxt project!

I believe you're right where a consumer could only be a separate Worker, or be HTTP pull based.

With Pages, currently you can only add a producer binding.
![image](https://github.com/nuxt-hub/core/assets/20425781/02dee88f-e994-4174-925f-96d24883c06e)

**Comment by valtlfelipe** [comment]
I'm sorry if it's not the right place to ask, but I would like to publish a message on a CF queue from a hub project. But how would that be possible if we haven't the binding in hub yet, and I don't have access to the wrangler file to create de binding manually? Anyone had to do this already?

Also, if I may contribute with my feedback, Pages currently supports producers, and there is an HTTP API for pulling and ack messages. Simple composables, similar to the KV, would be really nice and simple to use.

**Comment by Gerbuuun** [comment]
@valtlfelipe I use this util to publish to a Queue (based on other nuxthub utils)
```ts
import type { Queue } from '@cloudflare/workers-types/experimental';

interface Message {
  orderId: string;
  lang: string;
}
type OrderQueue = Queue<Message>;

let queue: OrderQueue;

export function useQueue() {
  if (queue)
    return queue;

  // @ts-expect-error globalThis.__env__ is not defined
  // eslint-disable-next-line node/prefer-global/process
  const binding = process.env.QUEUE || globalThis.__env__?.QUEUE || globalThis.QUEUE;

  if (binding) {
    queue = binding as OrderQueue;
    return queue;
  }

  throw createError('Cloudflare Queue binding not found');
}
```
You should add the Queue binding to the worker in the cloudflare dashboard manually.

**Comment by valtlfelipe** [comment]
> @valtlfelipe I use this util to publish to a Queue (based on other nuxthub utils)
> [...]
> You should add the Queue binding to the worker in the cloudflare dashboard manually.

Nice workaround. Thanks!

**Comment by alexcroox** [comment]
This would be great thanks. I need to process 1000s of lines from a CSV and process through hubAI() in parallel using a queue system

**Comment by runarlevi** [comment]
@alexcroox so you did manage to get it working this way?

**Comment by alexcroox** [comment]
No I don’t know the internals of Nuxt/nitro to be able to modify either listening for fetch or queue message type in the cloudflare adaptor.

**Comment by runarlevi** [comment]
I see. Im hoping its possible though, it would be nice to not have to create a separate project for the consumer worker!

**Comment by RihanArfan** [comment]
It's now possible to add additional bindings to NuxtHub projects, including the queue binding (though it's unavailable during local development) https://hub.nuxt.com/changelog/observability-additional-bindings

We are planning on introducing `hubQueue()` too in the coming weeks.

> CF sends a special "queue" orign rather than "fetch" so it's just a matter of handling that instead.

You can access this using the Nitro hook `cloudflare:queue`. https://github.com/nitrojs/nitro/blob/v3/src/presets/cloudflare/runtime/_module-handler.ts#L72-L82

https://nuxt.com/docs/guide/going-further/hooks#server-hooks-runtime

**Comment by bgervan** [comment]
Is there any update? Or is there any manual workaround to make it work locally too?

**Comment by RihanArfan** [comment]
With NuxtHub v0.10, you can now directly add both Queue producer and consumer bindings as you now manage your own wrangler.jsonc.

```jsonc
// wrangler.jsonc
{
  "$schema": "./node_modules/wrangler/config-schema.json",
  "queues": {
    "producers": [
      {
        "queue": "my-queue",
        "binding": "MY_QUEUE"
      }
    ],
    "consumers": [
      {
        "queue": "my-queue",
        "max_batch_size": 10,
        "max_batch_timeout": 30,
        "max_retries": 10,
        "dead_letter_queue": "my-queue-dlq"
      }
    ]
  }
}
```

From Nuxt you can publish/send to the queue using the binding:
```ts
export default defineEventHandler((event) => {
  const { cloudflare } = event.context
  const queue = await cloudflare.env.MY_QUEUE

  const url = getRequestURL(event)
  let log = {
    url: request.url,
  };
  await queue.send(log);

  return { success: true }
})
```

And you can consume the queue from your Nuxt app using the Nitro hook `cloudflare:queue` https://nitro.build/guide/plugins
```ts
export default defineNitroPlugin((nitro) => {
  nitro.hooks.hook("cloudflare:queue", async ({ batch, event, env, context }) => {
    // Code to run when receiving a queue message
  });
})
```

https://github.com/nitrojs/nitro/blob/9617490c7b239722d6eb4d37be1a8e5c41b93725/src/presets/cloudflare/runtime/_module-handler.ts#L72-L82
